src/main/java/com/microsoft/graph/content/BatchRequestContent.java

```diff
@@ -0,0 +1,283 @@
+        return this.requestAdapter.convertToNativeRequestAsync(requestInformation).thenCompose(request -> {
+            BatchRequestStep requestStep = new BatchRequestStep(requestId, (Request) request);
+            this.batchRequestSteps.put(requestId, requestStep);
+            return CompletableFuture.completedFuture(requestId);
+        });
+    }
+    /**
+     * Removes a batch request step from the batch request.
+     * @param requestId The request id of the request to remove.
+     * @return True if the request was removed, false otherwise.
+     */
+    public boolean removeBatchRequestStepWithId(@Nonnull String requestId) {
+        if(requestId.isEmpty()) {
+            throw new IllegalArgumentException(ErrorConstants.Messages.NULL_PARAMETER + "requestId");
+        }
+        boolean isRemoved = false;
+        if(this.batchRequestSteps.containsKey(requestId)) {
+            this.batchRequestSteps.remove(requestId);
+            isRemoved = true;
+            for (BatchRequestStep requestStep : this.batchRequestSteps.values()) {
+                if(requestStep.getDependsOn() != null) {
+                    requestStep.removeDependsOnId(requestId);
+                }
+            }
+        }
+        return isRemoved;
+    }
+
+    /**
+     * Builds a BatchRequestContent object from failed requests.
+     * @param responseStatusCodes The response status codes of the failed requests.
+     * @return The BatchRequestContent object.
+     */
+    @Nonnull
+    public BatchRequestContent newBatchWithFailedRequests (@Nonnull Map<String, Integer> responseStatusCodes) {
+        BatchRequestContent request = new BatchRequestContent(this.requestAdapter, new ArrayList<>());
+        responseStatusCodes.forEach((key, value) -> {
+            if(this.batchRequestSteps.containsKey(key) && !BatchResponseContent.isSuccessStatusCode(value)) {
+                request.addBatchRequestStep(this.batchRequestSteps.get(key).getRequest());
+            }
+        });
+        return request;
+    }
+    /**
+     * Builds the json content of the batch request.
+     * @return The json content of the batch request as an InputStream.
+     */
+    @Nonnull
+    public CompletableFuture<InputStream> getBatchRequestContentAsync() {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        JsonWriter writer = new JsonWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+        try {
+            writer.setIndent("  ");
+            writer.beginObject();
+            writer.name(CoreConstants.BatchRequest.REQUESTS);
+            writer.beginArray();
+            for (BatchRequestStep requestStep : this.batchRequestSteps.values()) {
+                writeBatchRequestStepAsync(requestStep, writer);
+            }
+            writer.endArray();
+            writer.endObject();
+            writer.flush();
+
+            ByteArrayInputStream stream = new ByteArrayInputStream(outputStream.toByteArray());
```

Comment:
use pipes instead to avoid memory pressure.
https://stackoverflow.com/a/23874232/3808675
Since we're already in a completable future and most likely deferred to a thread pool, you don't need to the writing in a separate thread like they do in the examples.

Reply:
Please review the updated code block. Thanks!